[docs] Make walkthrough and starting Ray materials clear (#7099)
ray-project/ray: doc/source/memory-management.rst

For Ray initialization on a single node, consider setting the following fields:

.. code-block:: python

  ray.init(
      memory=2000 * 1024 * 1024,
      object_store_memory=200 * 1024 * 1024,
      driver_object_store_memory=100 * 1024 * 1024)

For Ray usage on a cluster, consider setting the following fields on both the command line and in your Python script:

.. code-block:: bash

  # On the head node
  ray start --head --redis-port=6379 \
      --object-store-memory=200000000 \
      --memory 200000000 \
      --num-cpus=1

  # On the worker node
  ray start --object-store-memory=200000000 \
      --memory 200000000 \
      --num-cpus=1 \
      --address=$RAY_HEAD_ADDRESS:6379

.. code-block:: python

  # In your Python script connecting to Ray:
  ray.init(
      address="localhost:6379",
      driver_object_store_memory=100 * 1024 * 1024
  )

For any custom remote method or actor, you can try setting requirements as follows:

.. code-block:: python

  @ray.remote(
      memory=2000 * 1024 * 1024,

>       object_store_memory=200 * 1024 * 1024)

Author: cc @ericl

Author: unfortunately........
```
In [1]: import ray
ra
In [2]: @ray.remote(
   ...:     memory=2000 * 1024 * 1024,
   ...:     object_store_memory=200 * 1024 * 1024)
   ...: def hi():
   ...:     return 1
   ...:
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-2-a5a00aed0ec9> in <module>
      1 @ray.remote(
      2     memory=2000 * 1024 * 1024,
----> 3     object_store_memory=200 * 1024 * 1024)
      4 def hi():
      5     return 1

~/miniconda3/lib/python3.7/site-packages/ray/worker.py in decorator(function_or_class)
   1695                 function_or_class, num_cpus, num_gpus, memory,
   1696                 object_store_memory, resources, num_return_vals, max_calls,
-> 1697                 max_retries)
   1698
   1699         if inspect.isclass(function_or_class):

~/miniconda3/lib/python3.7/site-packages/ray/remote_function.py in __init__(self, function, num_cpus, num_gpus, memory, object_store_memory, resources, num_return_vals, max_calls, max_retries)
     70         if object_store_memory is not None:
     71             raise NotImplementedError(
---> 72                 "setting object_store_memory is not implemented for tasks")
     73         self._object_store_memory = None
     74         self._resources = resources

NotImplementedError: setting object_store_memory is not implemented for tasks
```

Reviewer: It is really hard to read memory without explit units.
Maybe we can use something like [bitmath](https://github.com/tbielawa/bitmath):

```
 @ray.remote(
      memory = GB(2),
      object_store_memory=  MB(200)
...

```
This will make all of the code self-documenting.

Or?

```
 @ray.remote(
      memory_bytes=2097152000,
      object_store_memory_bytes=209715200)
```

This page describes how memory management works in Ray, and how you can set memory quotas to ensure memory-intensive applications run predictably and reliably.

Summary
-------

You can set memory quotas to ensure your application runs predictably on any Ray cluster configuration. If you're not sure, you can start with a conservative default configuration like the following and see if any limits are hit.

For Ray initialization on a single node, consider setting the following fields:

.. code-block:: python

  ray.init(
      memory=2000 * 1024 * 1024,
      object_store_memory=200 * 1024 * 1024,
      driver_object_store_memory=100 * 1024 * 1024)

For Ray usage on a cluster, consider setting the following fields on both the command line and in your Python script:

.. code-block:: bash

  # On the head node
  ray start --head --redis-port=6379 \

>       --object-store-memory=200000000 \

Reviewer 2: The memory limits here don't match in `ray.init()` above which is a bit confusing. Would stick with GiB over GB in both places.

Author: ```suggestion
      --object-store-memory=200 * 1024 * 1024 \
```

Reviewer: That's why I like standard lib/reference like this that new users can refer to:
https://bitmath.readthedocs.io/en/latest/classes.html

when they see:

`ray start --head --redis-port=6379 --object-store-memory=bitmath.MB(200)`

>       --memory 200000000 \

Author: ```suggestion
      --memory 2000 * 1024 * 1024 \
```

      --num-cpus=1

  # On the worker node
  ray start --object-store-memory=200000000 \
      --memory 200000000 \
      --num-cpus=1 \
      --address=$RAY_HEAD_ADDRESS:6379

.. code-block:: python

  # In your Python script connecting to Ray:
  ray.init(

>       address="localhost:6379",

Reviewer 2: ```suggestion
      address="auto",
```

Reviewer 2: Maybe with a comment saying you can set it to `host:port` if not using defaults

      driver_object_store_memory=100 * 1024 * 1024
  )

> For any custom remote method or actor, you can try setting requirements as follows:

Reviewer 2: ```suggestion
For any custom remote method or actor, you can set requirements as follows:
```

.. code-block:: python

  @ray.remote(
      memory=2000 * 1024 * 1024,

>       object_store_memory=200 * 1024 * 1024)

Author: ```suggestion
  )
```

For Ray usage on a cluster, consider setting the following fields on both the command line and in your Python script:

.. tip:: 200 * 1024 * 1024 bytes is 200 MiB. Use double parentheses to evaluate math in Bash: ``$((200 * 1024 * 1024))``.

.. code-block:: bash

  # On the head node
  ray start --head --redis-port=6379 \

>       --object-store-memory=$((200 * 1024 * 1024)) \

Reviewer 2: beautiful
